File: cj-sync-worker/shipment_risk.py
"""Pure shipment-risk scoring for the Commerce OS operations bridge.

The scorer never contacts customers, pays CJ, cancels an order, or issues a
refund.  It translates reconciled Shopify + CJ evidence into an owner-facing
priority and a plain-English next action.
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Any
from zoneinfo import ZoneInfo


ISRAEL_TZ = ZoneInfo("Asia/Jerusalem")
CJ_TZ = ZoneInfo("Asia/Shanghai")
ISRAEL_WEEKEND = {4, 5}  # Friday and Saturday

SEVERITY_WEIGHT = {
    "MONITORING": 0,
    "MEDIUM": 1,
    "HIGH": 2,
    "CRITICAL": 3,
}

EXCEPTION_TERMS = (
    "exception",
    "failed delivery",
    "delivery failed",
    "undeliverable",
    "return to sender",
    "returned to sender",
    "shipment returned",
    "delivery failure",
)

LABEL_ONLY_TERMS = (
    "label created",
    "warehouse is processing",
    "shipment information received",
    "pre-shipment",
)

# Where the parcel physically is, read from CJ's own event remarks. Ordered by
# progress; the most advanced matching event wins, so a repeated "arrived at
# customs" scan after "released from customs" does not move the parcel back.
STAGE_RULES: tuple[tuple[str, int, str, tuple[str, ...]], ...] = (
    ("LABEL_CREATED", 1, "Label created, warehouse packing", ("label created", "warehouse is processing", "shipment information received", "pre-shipment")),
    ("CN_WAREHOUSE", 2, "Left the CJ warehouse (China)", ("china warehouse", "e-post china", "picked up", "arrived at sorting", "shenzhen", "send to hongkong", "sent to hongkong")),
    ("TO_HONG_KONG", 3, "In Hong Kong, waiting for flight", ("arrived in hongkong", "arrived at hongkong", "hong kong", "hongkong", "waiting for flight")),
    ("IN_AIR", 4, "Departed Hong Kong, flying to Israel", ("departed from hongkong", "departed from hong kong", "flight departed", "departed")),
    ("IL_AIRPORT", 5, "Arrived at TLV airport", ("tlv airport", "arrived at tlv", "arrived at destination airport", "arrived in israel")),
    ("IL_CUSTOMS", 6, "In Israeli customs", ("israel customs", "arrived at customs", "customs clearance")),
    ("IL_CUSTOMS_RELEASED", 7, "Released from customs", ("released from customs", "customs released", "cleared customs")),
    ("IL_LAST_MILE", 8, "With the Israeli courier, heading to pickup point", ("prepared to be sent to pickup point", "sent to pickup point", "out for delivery", "in transit to pickup", "handed to local carrier", "last mile")),
    ("IL_READY_FOR_PICKUP", 9, "Waiting at the pickup point", ("ready for pickup", "arrived at pickup point", "available for pickup", "awaiting collection")),
    ("DELIVERED", 10, "Delivered", ("delivered", "collected by recipient", "picked up by recipient")),
)
ISRAEL_STAGES = {"IL_AIRPORT", "IL_CUSTOMS", "IL_CUSTOMS_RELEASED", "IL_LAST_MILE", "IL_READY_FOR_PICKUP"}


def tracking_stage(routes: list[dict[str, Any]]) -> tuple[str, int, str]:
    """Return (stage, progress, label) for the most advanced CJ event."""
    best = ("UNKNOWN", 0, "No carrier scan yet")
    for route in routes:
        remark = str(route.get("remark") or "").strip().lower()
        if not remark:
            continue
        for stage, progress, label, terms in STAGE_RULES:
            if any(term in remark for term in terms) and progress > best[1]:
                best = (stage, progress, label)
    return best


def parse_moment(value: Any, *, naive_tz: ZoneInfo = CJ_TZ) -> datetime | None:
    text = str(value or "").strip()
    if not text:
        return None
    try:
        parsed = datetime.fromisoformat(text.replace("Z", "+00:00"))
    except ValueError:
        return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=naive_tz)
    return parsed.astimezone(timezone.utc)


def business_days_between(start: datetime | None, end: datetime) -> int:
    """Count completed Israeli business dates after ``start`` through ``end``."""
    if start is None:
        return 0
    start_date = start.astimezone(ISRAEL_TZ).date()
    end_date = end.astimezone(ISRAEL_TZ).date()
    if end_date <= start_date:
        return 0
    cursor = start_date + timedelta(days=1)
    total = 0
    while cursor <= end_date:
        if cursor.weekday() not in ISRAEL_WEEKEND:
            total += 1
        cursor += timedelta(days=1)
    return total


def elapsed_days(start: datetime | None, end: datetime) -> int:
    if start is None:
        return 0
    return max(0, int((end - start).total_seconds() // 86_400))


def _route_moments(routes: list[dict[str, Any]]) -> list[datetime]:
    return [moment for route in routes if (moment := parse_moment(route.get("acceptTime")))]


def _has_physical_movement(routes: list[dict[str, Any]]) -> bool:
    for route in routes:
        remark = str(route.get("remark") or "").strip().lower()
        if remark and not any(term in remark for term in LABEL_ONLY_TERMS):
            return True
    return False


def _has_exception(tracking_status: str, routes: list[dict[str, Any]]) -> bool:
    evidence = " ".join(
        [tracking_status, *[str(route.get("remark") or "") for route in routes]]
    ).lower()
    return any(term in evidence for term in EXCEPTION_TERMS)


def _signal(
    code: str,
    severity: str,
    label: str,
    action: str,
    contact: str,
) -> dict[str, str]:
    return {
        "code": code,
        "severity": severity,
        "label": label,
        "action": action,
        "contact": contact,
    }


def score_shipment(record: dict[str, Any], *, now: datetime | None = None) -> dict[str, Any]:
    now = (now or datetime.now(timezone.utc)).astimezone(timezone.utc)
    order_created = parse_moment(record.get("orderCreatedAt"), naive_tz=ISRAEL_TZ)
    cj_created = parse_moment(record.get("cjCreatedAt"))
    cj_paid = parse_moment(record.get("cjPaymentAt"))
    out_warehouse = parse_moment(record.get("outWarehouseAt"))
    routes = list(record.get("trackingRoutes") or [])
    moments = _route_moments(routes)
    label_at = min(moments) if moments else parse_moment(record.get("trackingCreatedAt"))
    latest_tracking_at = max(moments) if moments else None
    order_business_days = business_days_between(order_created, now)
    label_business_days = business_days_between(label_at, now)
    inactive_days = elapsed_days(latest_tracking_at, now)
    cj_status = str(record.get("cjStatus") or "UNKNOWN").strip().upper()
    cj_sub_status = str(record.get("cjSubStatus") or "").strip().upper()
    tracking_status = str(record.get("trackingStatus") or "").strip()
    tracking_present = bool(record.get("trackingPresent"))
    shopify_fulfillment = str(record.get("shopifyFulfillmentStatus") or "").upper()
    shopify_financial = str(record.get("shopifyFinancialStatus") or "").upper()
    shopify_risk = str(record.get("shopifyRiskRecommendation") or "NONE").upper()
    sale_count = int(record.get("saleTransactionCount") or 0)
    duplicate_cj = bool(record.get("duplicateCjRecords"))
    stage, stage_progress, stage_label = tracking_stage(routes)
    latest_route = max(routes, key=lambda route: parse_moment(route.get("acceptTime")) or datetime.min.replace(tzinfo=timezone.utc)) if moments else None
    latest_remark = str((latest_route or {}).get("remark") or "").strip() or None
    in_israel = stage in ISRAEL_STAGES
    last_mile = stage in {"IL_LAST_MILE", "IL_READY_FOR_PICKUP"}
    delivered = "DELIVERED" in {cj_status, tracking_status.upper(), shopify_fulfillment} or stage == "DELIVERED"
    signals: list[dict[str, str]] = []

    if duplicate_cj:
        signals.append(_signal(
            "DUPLICATE_CJ_RECORDS", "CRITICAL", "More than one CJ record",
            "Resolve the duplicate CJ records before any further fulfillment action.",
            "Internal + CJ",
        ))

    if shopify_financial in {"REFUNDED", "VOIDED"} and cj_status not in {"CANCELLED", "CLOSED", "UNKNOWN"}:
        signals.append(_signal(
            "REFUND_FULFILLMENT_CONFLICT", "CRITICAL", "Refund and fulfillment conflict",
            "Stop and reconcile Shopify payment state with CJ before the parcel progresses.",
            "Internal + CJ",
        ))

    if _has_exception(tracking_status, routes):
        signals.append(_signal(
            "CARRIER_EXCEPTION", "CRITICAL", "Carrier exception",
            "Escalate to CJ today and prepare a concrete customer solution for approval.",
            "CJ + Customer",
        ))

    payment_required = cj_status in {"CREATED", "IN_CART", "UNPAID"} and not cj_paid
    if payment_required:
        signals.append(_signal(
            "CJ_PAYMENT_REQUIRED", "HIGH", "CJ payment or confirmation required",
            "Confirm CJ payment and processing status today. Do not promise dispatch until CJ confirms.",
            "CJ",
        ))

    lifecycle_start = cj_paid or cj_created or order_created
    lifecycle_business_days = business_days_between(lifecycle_start, now)
    if not tracking_present and not payment_required and not delivered:
        if lifecycle_business_days >= 5:
            signals.append(_signal(
                "TRACKING_NOT_FOUND_CRITICAL", "CRITICAL", f"No tracking after {lifecycle_business_days} business days",
                "Open a CJ escalation now and request proof of warehouse handoff.",
                "CJ",
            ))
        elif lifecycle_business_days >= 2:
            signals.append(_signal(
                "TRACKING_NOT_FOUND_WARNING", "HIGH", f"No tracking after {lifecycle_business_days} business days",
                "Ask CJ for the expected tracking and warehouse handoff time.",
                "CJ",
            ))

    # CJ's warehouse routinely takes 4-6 business days between the label and
    # the first physical scan on this lane; alerting at day 2 painted half the
    # board red for parcels that were simply being packed.
    physical_movement = bool(out_warehouse) or _has_physical_movement(routes)
    if tracking_present and not physical_movement and not delivered:
        if label_business_days >= 8:
            signals.append(_signal(
                "LABEL_NO_PICKUP_CRITICAL", "CRITICAL", f"Label without pickup for {label_business_days} business days",
                "Escalate to CJ now and request a physical pickup scan or a replacement plan.",
                "CJ",
            ))
        elif label_business_days >= 5:
            signals.append(_signal(
                "LABEL_NO_PICKUP_WARNING", "HIGH", f"Label without pickup for {label_business_days} business days",
                "Ask CJ to confirm when the parcel will physically leave the warehouse.",
                "CJ",
            ))

    # A parcel waiting at an Israeli pickup point legitimately shows no scans
    # for days; only silence before the last mile is a carrier problem.
    stale_threshold = 6 if last_mile else 3
    if tracking_present and physical_movement and not delivered and inactive_days >= stale_threshold:
        signals.append(_signal(
            "TRACKING_STALE", "MEDIUM", f"No tracking update for {inactive_days} days ({stage_label})",
            "Request a movement update and prepare a customer update if the carrier cannot confirm progress.",
            "CJ",
        ))

    # The day count alone is not the emergency; a parcel that cleared Israeli
    # customs on day 17 needs no remedy. What needs one is a parcel that is
    # both late and silent, or late beyond any normal transit.
    moving = latest_tracking_at is not None and inactive_days < 4
    if not delivered and (order_business_days >= 22 or (order_business_days >= 14 and not moving and not in_israel)):
        signals.append(_signal(
            "DAY_14_SOLUTION_DUE", "CRITICAL",
            f"Not delivered after {order_business_days} business days, no movement for {inactive_days} days" if latest_tracking_at else f"Not delivered after {order_business_days} business days",
            "Choose a concrete remedy for the customer; do not send another waiting-only reply.",
            "Customer + CJ",
        ))
    elif not delivered and order_business_days >= 14:
        signals.append(_signal(
            "DAY_14_MOVING", "MEDIUM", f"Day {order_business_days}: {stage_label}",
            "Late but moving; the proactive delivery update goes out automatically. Watch for the next scan.",
            "Monitor",
        ))
    elif not delivered and order_business_days >= 10 and not moving and not in_israel:
        signals.append(_signal(
            "DAY_10_CUSTOMER_UPDATE", "HIGH", f"Business day {order_business_days}, no movement for {inactive_days} days",
            "Ask CJ for a movement update; the proactive customer update goes out automatically.",
            "Customer",
        ))

    if shopify_risk == "CANCEL":
        signals.append(_signal(
            "SHOPIFY_HIGH_RISK", "HIGH", "Shopify recommends urgent risk review",
            "Review the order manually today. Do not cancel automatically.",
            "Internal",
        ))
    elif shopify_risk == "INVESTIGATE":
        signals.append(_signal(
            "SHOPIFY_MEDIUM_RISK", "MEDIUM", "Shopify recommends manual review",
            "Review the Shopify risk evidence before taking any order action.",
            "Internal",
        ))

    if sale_count > 1:
        signals.append(_signal(
            "MULTIPLE_SALE_TRANSACTIONS", "MEDIUM", f"{sale_count} successful SALE transactions",
            "Review Shopify transactions and AfterSell before explaining the total charge to the customer.",
            "Internal",
        ))

    signals.sort(key=lambda item: SEVERITY_WEIGHT[item["severity"]], reverse=True)
    primary = signals[0] if signals else _signal(
        "MONITORING", "MONITORING",
        (stage_label if stage != "UNKNOWN" else "Moving normally") if tracking_present else "Within the normal preparation window",
        "No action needed. Keep monitoring the next verified milestone.",
        "Monitor",
    )
    return {
        **record,
        "severity": primary["severity"],
        "primarySignal": primary["code"],
        "statusLabel": primary["label"],
        "doNow": primary["action"],
        "contactTarget": primary["contact"],
        "signals": signals,
        "isActionable": bool(signals),
        "orderBusinessDays": order_business_days,
        "labelBusinessDays": label_business_days,
        "inactiveDays": inactive_days,
        "delivered": delivered,
        "sourceAgreement": str(
            record.get("sourceAgreement")
            or ("CONFLICT" if duplicate_cj else "VERIFIED")
        ),
        "afterSellLikely": sale_count > 1,
        "latestTrackingAt": latest_tracking_at.isoformat() if latest_tracking_at else None,
        "trackingStage": stage if tracking_present else None,
        "trackingStageLabel": stage_label if tracking_present else None,
        "latestRemark": latest_remark,
        "inIsrael": in_israel,
        "trackingCreatedAt": label_at.isoformat() if label_at else None,
        "cjSubStatus": cj_sub_status or None,
    }

File: app/cloudflare-pilot/src/lib/shipment-risk.ts
import { normalizeCjTracking, type CjTrackingRoute, type ShipmentStatus } from "./cj-tracking.js";

/**
 * Re-scores a monitor record from the carrier's own events.
 *
 * The Python monitor scores by the calendar: "not delivered after 14 business
 * days" was CRITICAL even for a parcel released from Israeli customs that
 * morning, and a label two days old was already HIGH. That painted most of the
 * board red and hid the handful of orders that actually need a person. The
 * monitor sends the raw CJ routes with every record, so the Worker can rescore
 * here, on ingest, without depending on a Railway deploy.
 */
export interface ShipmentSignal {
  code: string;
  severity: "MONITORING" | "MEDIUM" | "HIGH" | "CRITICAL";
  label: string;
  action: string;
  contact: string;
}

const WEIGHT: Record<ShipmentSignal["severity"], number> = { MONITORING: 0, MEDIUM: 1, HIGH: 2, CRITICAL: 3 };

/** Signals the Python scorer derives from day counts alone; replaced below. */
const CALENDAR_SIGNALS = new Set([
  "DAY_14_SOLUTION_DUE", "DAY_14_MOVING", "DAY_10_CUSTOMER_UPDATE",
  "LABEL_NO_PICKUP_CRITICAL", "LABEL_NO_PICKUP_WARNING", "TRACKING_STALE",
]);

const STAGE_LABEL: Record<string, string> = {
  UNKNOWN: "Moving normally",
  LABEL_CREATED: "Label created, warehouse packing",
  CN_WAREHOUSE: "Left the CJ warehouse (China)",
  TO_HONG_KONG: "In Hong Kong, waiting for flight",
  IN_AIR: "Departed Hong Kong, flying to Israel",
  IL_AIRPORT: "Arrived at TLV airport",
  IL_CUSTOMS: "In Israeli customs",
  IL_CUSTOMS_RELEASED: "Released from customs",
  IL_LAST_MILE: "With the Israeli courier, heading to pickup point",
  IL_READY_FOR_PICKUP: "Waiting at the pickup point",
  DELIVERED: "Delivered",
  EXCEPTION: "Carrier exception",
};

export interface RescoreInput {
  severity: string;
  primarySignal: string | null;
  statusLabel: string | null;
  doNow: string | null;
  contactTarget: string | null;
  signals: ShipmentSignal[];
  delivered: boolean;
  trackingPresent?: boolean;
  trackingLast4?: string | null;
  trackingNumber?: string | null;
  trackingStatus?: string | null;
  cjMailNo?: string | null;
  trackingRoutes?: CjTrackingRoute[] | null;
  orderBusinessDays: number;
  labelBusinessDays: number;
  outWarehouseAt?: string | null;
}

export interface RescoreResult {
  severity: ShipmentSignal["severity"];
  primarySignal: string;
  statusLabel: string;
  doNow: string;
  contactTarget: string;
  signals: ShipmentSignal[];
  delivered: boolean;
  inactiveDays: number;
  trackingStage: string | null;
  latestRemark: string | null;
  latestTrackingAt: string | null;
  status: ShipmentStatus | null;
}

export function rescoreShipment(input: RescoreInput, now: Date = new Date(), precomputed: ShipmentStatus | null = null): RescoreResult {
  const routes = Array.isArray(input.trackingRoutes) ? input.trackingRoutes.filter(r => r && typeof r === "object") : [];
  const trackingPresent = Boolean(input.trackingPresent ?? (input.trackingNumber || input.trackingLast4));
  const status = precomputed ? precomputed : trackingPresent
    ? normalizeCjTracking({ trackingNumber: input.trackingNumber || input.trackingLast4 || "?", trackingStatus: input.trackingStatus || null, cjMailNo: input.cjMailNo || null, routes }, now)
    : null;
  const stage = status?.stage || null;
  const stageLabel = STAGE_LABEL[stage || "UNKNOWN"] || "Moving normally";
  const inIsrael = Boolean(status?.inIsrael);
  const lastMile = stage === "IL_LAST_MILE" || stage === "IL_READY_FOR_PICKUP";
  const delivered = Boolean(input.delivered) || Boolean(status?.delivered);
  const inactiveDays = status?.inactiveDays ?? 0;
  const hasScan = Boolean(status?.latestAt);
  const physicalMovement = Boolean(input.outWarehouseAt) || (hasScan && stage !== "LABEL_CREATED" && stage !== "UNKNOWN");
  const days = Math.max(0, Math.floor(Number(input.orderBusinessDays) || 0));
  const labelDays = Math.max(0, Math.floor(Number(input.labelBusinessDays) || 0));

  const signals: ShipmentSignal[] = (input.signals || []).filter(s => s && !CALENDAR_SIGNALS.has(String(s.code)));
  const push = (code: string, severity: ShipmentSignal["severity"], label: string, action: string, contact: string) =>
    signals.push({ code, severity, label, action, contact });

  if (!delivered && trackingPresent) {
    // CJ's warehouse takes 4-6 business days between the label and the first
    // physical scan on this lane; alerting at day 2 was noise.
    if (!physicalMovement) {
      if (labelDays >= 8) push("LABEL_NO_PICKUP_CRITICAL", "CRITICAL", `Label without pickup for ${labelDays} business days`, "Escalate to CJ now and request a physical pickup scan or a replacement plan.", "CJ");
      else if (labelDays >= 5) push("LABEL_NO_PICKUP_WARNING", "HIGH", `Label without pickup for ${labelDays} business days`, "Ask CJ to confirm when the parcel will physically leave the warehouse.", "CJ");
    }
    // A parcel waiting at an Israeli pickup point legitimately shows no scans
    // for days; silence before the last mile is a carrier problem.
    const staleThreshold = lastMile ? 6 : 3;
    if (physicalMovement && hasScan && inactiveDays >= staleThreshold) {
      push("TRACKING_STALE", "MEDIUM", `No tracking update for ${inactiveDays} days (${stageLabel})`, "Request a movement update and prepare a customer update if the carrier cannot confirm progress.", "CJ");
    }
  }

  // The day count alone is not the emergency; what needs a remedy is a parcel
  // that is both late and silent, or late beyond any normal transit.
  const moving = hasScan && inactiveDays < 4;
  if (!delivered && (days >= 22 || (days >= 14 && !moving && !inIsrael))) {
    push("DAY_14_SOLUTION_DUE", "CRITICAL",
      hasScan ? `Not delivered after ${days} business days, no movement for ${inactiveDays} days` : `Not delivered after ${days} business days`,
      "Choose a concrete remedy for the customer; do not send another waiting-only reply.", "Customer + CJ");
  } else if (!delivered && days >= 14) {
    push("DAY_14_MOVING", "MEDIUM", `Day ${days}: ${stageLabel}`, "Late but moving; the proactive delivery update goes out automatically. Watch for the next scan.", "Monitor");
  } else if (!delivered && days >= 10 && !moving && !inIsrael && trackingPresent) {
    push("DAY_10_CUSTOMER_UPDATE", "HIGH", `Business day ${days}, no movement for ${inactiveDays} days`, "Ask CJ for a movement update; the proactive customer update goes out automatically.", "Customer");
  }

  if (status?.exception && !signals.some(s => s.code === "CARRIER_EXCEPTION")) {
    push("CARRIER_EXCEPTION", "CRITICAL", "Carrier exception", "Escalate to CJ today and prepare a concrete customer solution for approval.", "CJ + Customer");
  }

  signals.sort((a, b) => (WEIGHT[b.severity] ?? 0) - (WEIGHT[a.severity] ?? 0));
  const primary: ShipmentSignal = signals[0] || {
    code: "MONITORING", severity: "MONITORING",
    label: delivered ? "Delivered" : trackingPresent ? stageLabel : "Within the normal preparation window",
    action: "No action needed. Keep monitoring the next verified milestone.", contact: "Monitor",
  };
  return {
    severity: primary.severity,
    primarySignal: primary.code,
    statusLabel: primary.label,
    doNow: primary.action,
    contactTarget: primary.contact,
    signals,
    delivered,
    inactiveDays,
    trackingStage: stage,
    latestRemark: status?.latestRemark || null,
    latestTrackingAt: status?.latestAt || null,
    status,
  };
}
